### ISSUE TITLE
[Feature Request] Support for Chrome's built in tab-groups feature - now available from API?

### ISSUE THREAD

**Issue by RedMoikle** [gilgold/tabox#6]
Hi there, I am a Python dev who does workflow automation and data manipulation. I love that this extension exists, and it is almost exactly what I have been itching to build myself (once I get round to learning javascript at least).

I'll use the words "Chrome group" to refer to Google's own built in system, and "Tabox group" to refer to your system.

I am not a JS dev, so excuse me if you already know this, and/or there is another roadblock I am not aware of, but I have read some of your comments about wanting to support Chrome's own built in tab-groups feature. This would be an absolute gamechanger for me. The Chrome groups are essential for me to keep things organised, and having tabs on different subjects keeps things visually separate (and minimisable!)

**Creating new Chrome groups with Tabox contents**
I believe what you may be looking for has been added to the API, and is here: 
https://developer.chrome.com/docs/extensions/reference/tabs/#method-group
This should create a new chrome group (or select an existing one), and you can then give it a list of tabIds to add to the group. You could have an optional parameter to automatically put Tabox groups into a new Chrome group with the same name after they are loaded.

**Creating a tabox group from chrome group contents**
https://developer.chrome.com/docs/extensions/reference/tabs/#type-Tab
The groupId property of the Tab object should also let you access the Chrome group that the tab is in, and could potentially be used to create a Tabox group from all of the tabs in a selected Chrome group without needing to select them all individually. 
This second feature isn't quite as important, since it isn't hard to shift-select a Chrome group's contents, but it would be a nice quality of life addition.

Again, apologies if I am barking up the wrong tree or telling you things you already know, I am a total JS noob

Keep up the great work!

**Comment by RedMoikle** [comment]
https://developer.chrome.com/docs/extensions/reference/tabs/#method-query

The query method looks like it can also be used to list all tabs in the specified group.

some pseudocode - I am not going to submit this as a pull request, as I am not a JS Dev:

```
// near the top of your DOMContentLoaded Code - line 102/103 at time of writing

if (tabs.length > 1) {
      saveSelectedOnly = true;
      tabTypeWord.text('selected');
      highlighter.text(`${tabs.length} selected`);
      highlighted_note.fadeIn();

    // ---------- suggestion starts here ----------

    //you will probably need a different button/shortcut to trigger this, to preserve the original behavior when needed
    } else if saveChromeGroups {
      saveSelectedOnly = true;
      //you could probably do a for/in loop here to save multiple groups at once by selecting one tab from each chrome group
      let chromeGroupId= tabs[0].groupId;
      chrome.tabs.query({groupId=chromeGroupId}, function (chromeTabs) {
        // I am not sure how to handle modifying an external scope from a callback in JS so this will need to be changed by someone who knows JS
        tabs = chromeTabs; 
      });
      tabTypeWord.text('Chrome Group');
      highlighter.text(`${tabs.length} selected`);
      highlighted_note.fadeIn();

    // ---------- suggestion ends here ----------

    } else {
      saveSelectedOnly = false;
      tabTypeWord.text('open');
      highlighted_note.fadeOut();
    }
  });
```

**Comment by RedMoikle** [comment]
Could something like this work to load Tabox groups into a new chrome group?

```
// around line 186: added a parameter for group name and "useChromeGroups" option to toggle this behavior

  // Clicking on one of the saved tab groups
  $('#settings_body').on('click', '.settings_div', ({currentTarget}) => {
    
    // ---------- default group name as empty string means don't create a group ----------
    var groupName = ""; 
    // ----------------

    const setting_id = ...

    // ---------- if useChromeGroups setting is turned on, set groupname to match tabox name ---------- 
    if useChromeGroups {groupName = setting_id;} 
    // ---------------------

    settings_data.forEach((value) => {
      const single_setting_arr = ....
      
      // ---------- added groupname parameter----------
      if (single_setting_arr[0] == setting_id) openTabsWithUrls(single_setting_arr, groupName); 
      // ---------------

    });
  });

//--------------------------------------------------------
// around line 301: added groupName parameter, callback function to store created tabs in an array, then create group using tabs in this array.
function openTabsWithUrls(arr, groupName="") {
  // open a new Chrome window and load the saved URLs into tabs in it.

  // ---------- keep track of tabs created ---------- 
  var newTabs = []; 
  //---------------------

  arr.shift();
  if ($('#chkOpenNewWindow').is(':checked'))
  {
    chrome.windows.create({focused:true},(window) => {
      // create a new window
      ...
        // create the tab for the current url
        ...
        } else {
          const tab_properties = ...

          // ---------- callback adds tab to newTabs ----------
          setTimeout(() => chrome.tabs.create(tab_properties, function (tab) {newTabs.push(tab);}), 1); // ---------- callback adds tab to newTabs ----------
          // --------------------    

        }
      });
    });
  } else {
    // Use current window
    ....
        // create the tab for the current url
        ....
        } else {

          // ---------- callback adds tab to newTabs ----------
          setTimeout(() => chrome.tabs.create(tab_properties, function (tab) {newTabs.push(tab);}), 1);  
          // --------------------         
        }
      });
    });
  }

  // ---------- Create and fill group if a groupName was specified ----------      
  if groupName != "":
    // I haven't figured out how to name a tab group from the API yet, so for now I am just checking if the name parameter isn't empty.
    chrome.tabs.group((tabIds:newTabs));
  // --------------------
}
```

**Comment by gilgold** [comment]
Hey @RedMoikle ! Thank you for taking the time to dig into this!
I am currently working on various features regarding Chrome Groups.
I have been playing around with those APIs in the canary build of Chrome but had some issues with properly recreating chrome groups. I am still working on it and hope to have something ready soon. 

I appreciate the time you took to research the matter and suggesting improvements to Tabox!

**Comment by RedMoikle** [comment]
No worries!

I am glad to hear that you are working on it right now.
I'd like to mention that reading through your code taught me a little bit of JavaScript, so this has been useful for me!

**Comment by RedMoikle** [comment]
You seem to be getting there with this. The UI now recognises that the tabs were in a group, however they don't yet create a group when loading them again

**Comment by gilgold** [comment]
@RedMoikle Not sure what you mean, I tested it and when I save tabs that also include chrome groups, everything is saved correctly and when I click to open that tabox group it opens everything the way it was. 
If this is not what you are seeing, can you share some screenshots? 
Thanks!

**Comment by RedMoikle** [comment]
Test tabs and groups before I use tabox:
![image](https://user-images.githubusercontent.com/10142045/112151744-790d2200-8bd9-11eb-8f1d-5f24a40499cb.png)

I then used tabox to save all opened tabs and it does register that 2 groups are present:
![image](https://user-images.githubusercontent.com/10142045/112151852-9641f080-8bd9-11eb-9d92-368b2df39134.png)
![image](https://user-images.githubusercontent.com/10142045/112151893-a063ef00-8bd9-11eb-92a8-231e45fc857d.png)

I then close the window, open a new chrome window (at the new tab page) then try to re-open the saved tabs

However when I click on the entry in tabox, the tabs open but groups are not preserved:
![image](https://user-images.githubusercontent.com/10142045/112151992-ba9dcd00-8bd9-11eb-82d8-1eecd802f49c.png)

**Comment by gilgold** [comment]
I see, thank you for the screenshots!
I will look into this

**Comment by gilgold** [comment]
Looks like this is an issue only when using the "tabs will be added to this window" option instead of the open new window one.
I will create a new issue for this as it's not related to the chrome groups feature.
Thanks for reporting!

**Comment by gilgold** [comment]
@RedMoikle this should now be resolved in version 3.3.1 - it is live in the Chrome Web Store and should auto update in your browser soon.
